src/librustc_resolve/macros.rs

```diff
@@ -462,6 +479,12 @@ impl<'a, 'cl> Resolver<'a, 'cl> {
             return def;
         }
 
+        if kind == MacroKind::Attr && path.len() == 1 {
+            if let Some(ext) = self.unshadowable_attrs.get(&path[0].name) {
+                return Ok(ext.def());
+            }
+        }
```

Comment:
The proper solution, I think would be to bite the bullet and split macro namespace into three sub-namespaces, so that all fn-like macro resolutions are ignored when we are resolving an attribute and vice versa.

Reply:
In the meantime, I think, you can apply a more local hack and
- manually skip "bang" macros for `test` in `resolve_lexical_macro_path_segment` like it's done in https://github.com/petrochenkov/rust/commit/b2a9988791f7874fa7fad9a8454dd9e8748619bc for `cfg` and `thread_local`
- Add unshadowable_attrs as one more `WhereToResolve` layer between `MacroPrelude` and `BuiltinAttrs`.
- skip `resolve_legacy_scope` entirely if `kind == MacroKind::Attr && path[0].name == "test"`.

This would be forward-compatible with the proper solution.

Reply:
@kennytm 
I'm not happy about this as well, but cases like this continue to appear.
E.g. with strict rules built-in `#[cfg]`, `#[thread_local]` and `#[warn]` should all cause errors because they are shadowed by bang macros `cfg!(..)`, `thread_local { ... }` and [possibly `log::warn`](https://github.com/rust-lang/rust/issues/53205) respectively. And the question "why attributes and macros share a namespace at all" [always](https://github.com/rust-lang/rust/issues/53205#issuecomment-411552763) [arises](https://github.com/rust-lang/rust/issues/53583#issuecomment-415447800) when people see errors like this.

Reply:
My solution matches the current behavior of the compiler. If, in the future, we split the namespaces and want to allow shadowing it would be as simple as changing the [registration](https://github.com/rust-lang/rust/pull/53410/files#diff-6f54f41ca411f948b8597c829c4a9d36R74).

The more local hack you've mentioned seems a bit messier to clean up and would change the behavior of the attribute when my goal for this PR was to essentially just be a refactoring of `#[test]` and `#[bench]`.

Reply:
>My solution matches the current behavior of the compiler.

The issue is that the behavior *needs* to be changed and it needs to be changed soon, until freshly stabilized attribute macros get into beta.
That's why I'm suggesting to do it as a part of this PR, so I don't have to redo all this myself immediately after this PR lands :)

>If, in the future, we split the namespaces and want to allow shadowing it would be as simple as changing the registration.

Then `#[test]` will change meaning if it's shadowed by an attribute macro. We can avoid that future breaking change if we make it now while attribute macros are unstable.